crypto/fipsmodule/cpucap/internal.h

```diff
@@ -167,7 +167,7 @@ extern uint32_t OPENSSL_armcap_P;
 // |CRYPTO_is_ARMv8_AES_capable| and |CRYPTO_is_ARMv8_PMULL_capable|
 // for checking the support for AES and PMULL instructions, respectively.
 OPENSSL_INLINE int CRYPTO_is_NEON_capable(void) {
-#if defined(OPENSSL_STATIC_ARMCAP_NEON) || defined(__ARM_NEON)
```

Comment:
Why were the second parts of the checks removed? Aren't these possible compiler flags that can set the capability statically?

Reply:
According to [ACLE](https://developer.arm.com/documentation/ihi0053/d/?lang=en), `__ARM_NEON` is always set to 1 for AARCH64 platforms. As such, if we want `OPENSSL_armcap` to determine use of the C implementation, we can't check for this flag since it will always be set. We could use the `-march` flag when compiling to disable NEON, but since `OPENSSL_armcap` is supposed to determine processor capability at runtime I decided on this approach to reduce confusion.

`cpu_aarch64_apple.c` (as well as the counterpart files for other CPUs) still set `OPENSSL_armcap` to be the correct value for the processor in `OPENSSL_cpuid_setup()`, so removing the check doesn't disable NEON entirely, either.

This applies to `__ARM_FEATURE_AES` as well.

Reply:
Couldn't these values be used on ARM processors (32-bit)? We have users who build AWS-LC on them.

Reply:
[cpucap.c](https://github.com/aws/aws-lc/blob/main/crypto/fipsmodule/cpucap/cpucap.c#L53) detects if LC is built on 32-bit ARM and sets `OPENSSL_armcap` accordingly, so when we call a function like `CRYPTO_is_NEON_capable()`, even without a check for these values, it will correctly detect the CPU capabilities.

Reply:
I may be missing something, but I think if `OPENSSL_STATIC_ARMCAP` is defined but not `OPENSSL_STATIC_ARMCAP_NEON` due to an old compiler that defines `__ARM_NEON`, it will not reach the check for `OPENSSL_armcap` and will return 0.

Reply:
I changed the functions to no longer include the `ifdef` statements, similar to how the capability functions for x86 processors in the same file do it. Since `OPENSSL_armcap_P` is set correctly even when `OPENSSL_STATIC_ARMCAP` is defined, the functions should still return the right result.

Either way, a `BORINGSSL_DISPATCH_TEST` to ensure that this doesn't break anything will be introduced in a separate PR (which we should merge before this one)

Reply:
I think we'd also need to measure the performance in FIPS build with and without the ifdefs for static build (possibly for dynamic as well) when the STATIC_ARMCAPs are defined. Fetching the OPENSSL_armcap_P variable had some implications on the static FIPS build in x86 https://github.com/aws/aws-lc/pull/856.

Reply:
Performed a benchmark on both the dynamic and static builds comparing performance of AES on main to the PR branch:
* FIPS mode set
* `OPENSSL_STATIC_ARMCAP` and other static macros set
* Graviton2 on AL2

No performance differences between main & PR for either the dynamic nor static builds.

Full results can be seen here:
[aws-lc-dynamic-staticarmcap-aes_main_vs_nonhw_bm.csv](https://github.com/aws/aws-lc/files/12164501/aws-lc-dynamic-staticarmcap-aes_main_vs_nonhw_bm.csv)
[aws-lc-static-staticarmcap-aes_main_vs_nonhw_bm.csv](https://github.com/aws/aws-lc/files/12164502/aws-lc-static-staticarmcap-aes_main_vs_nonhw_bm.csv)